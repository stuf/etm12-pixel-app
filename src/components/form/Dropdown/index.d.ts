import { FunctionComponent } from 'react';
import { Value } from 'types';
import { IPalette } from 'core/models';

export interface Props {
  items: IPalette[];
  value: Value<number>;
}

export interface Component extends FunctionComponent<Props> {}
