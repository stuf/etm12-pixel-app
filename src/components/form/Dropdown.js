/* eslint no-unused-vars: [1, {"varsIgnorePattern": "[K|T]"}] */
import * as React from 'karet';
import * as U from 'karet.util';
import * as L from 'kefir.partial.lenses';

import * as M from 'common/meta';

export default function Dropdown({
  items,
  value,
  onChange = e => value.set(e.target.value),
  className,
}) {
  return (
    <div className={U.cns('formField', '--dropdown', className)}>
      {U.when(
        items,
        <select {...{ onChange }}>
          {U.thru(
            U.view(L.valueOr([]), items),
            U.mapElems((it, i) => (
              <option {...{ key: i, value: i }}>{M.nameIn(it)}</option>
            )),
          )}
        </select>,
      )}
    </div>
  );
}
