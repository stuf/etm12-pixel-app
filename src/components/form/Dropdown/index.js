/* eslint no-unused-vars: [1, {"varsIgnorePattern": "[K|T]"}] */
import * as React from 'karet';
import * as U from 'karet.util';

import * as T from './index.d';
import styles from './index.module.scss';

import * as M from 'common/meta';

/**
 * @param {T.Props} props
 * @return {T.Component}
 */
export default function Dropdown({ items, value }) {
  console.log({ items, value });
  return (
    <div className={styles.root}>
      <select onChange={e => value.set(e.target.value)}>
        {U.thru(
          items,
          U.mapElems((it, i) => (
            <option {...{ key: i, value: i }}>{M.nameIn(it)}</option>
          )),
        )}
      </select>
    </div>
  );
}
