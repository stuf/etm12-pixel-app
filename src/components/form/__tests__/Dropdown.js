import React from 'react';
import { mount, render } from 'enzyme';
import toJson from 'enzyme-to-json';

import Dropdown from 'components/form/Dropdown';

it('matches snapshot', () => {
  const xs = [{ name: 'top' }, { name: 'kek' }];

  const w1 = mount(<Dropdown />);
  const w2 = mount(<Dropdown items={xs} />);

  expect(toJson(w1)).toMatchSnapshot('empty dropdown');
  expect(toJson(w2)).toMatchSnapshot('non-empty dropdown');
});

it('allows setting a value to an atom', () => {
  const value = { set: jest.fn() };
  const xs = [{ name: 'top' }, { name: 'kek' }];

  const w = mount(<Dropdown items={xs} value={value} />);
  w.find('select').simulate('change', { target: { value: 123 } });

  expect(value.set).toHaveBeenCalled();
});
