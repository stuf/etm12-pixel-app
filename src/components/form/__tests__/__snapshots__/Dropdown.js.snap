// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot: empty dropdown 1`] = `
<Dropdown>
  <div
    className="formField --dropdown"
  />
</Dropdown>
`;

exports[`matches snapshot: non-empty dropdown 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "name": "top",
      },
      Object {
        "name": "kek",
      },
    ]
  }
>
  <div
    className="formField --dropdown"
  >
    <select
      onChange={[Function]}
    >
      <option
        key="0"
        value={0}
      >
        top
      </option>
      <option
        key="1"
        value={1}
      >
        kek
      </option>
    </select>
  </div>
</Dropdown>
`;
